### PR TITLE
Proposal for tenant retention

### PR DESCRIPTION
**Pull request by AllenzhLi**
This is a proposal / design doc for tenant retention data.

### CODE REVIEW

**Comment by pstibrany** [comment]
We typically use "runtime config" (overrides) for configuring per-tenant limits and options. Is there a good reason not to do that for tenant retention? If so, I'd suggest to include those reasons in the design document. Otherwise, I'd suggest that we use "overrides" for configuring tenant retention as well.

**Comment by AllenzhLi** [comment]
> We typically use "runtime config" (overrides) for configuring per-tenant limits and options. Is there a good reason not to do that for tenant retention? If so, I'd suggest to include those reasons in the design document. Otherwise, I'd suggest that we use "overrides" for configuring tenant retention as well. It would also simplify design – there will be no need to expose API for configuring or checking tenant retention.

Thank you for your reply.

The design is inspired by your tenant-deletion design.
 
The other reason to not use runtime config is that I does't find where can store the setting for every tenant.

**Comment by pstibrany** [comment]
> The other reason to not use runtime config is that I does't find where can store the setting for every tenant. Maybe I missing something.

Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.

File: docs/proposals/tenant_retention.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+- Author: [Allenzhli](https://github.com/Allenzhli)
+- Date: December 2020
+- Status: Proposed
+
+## Retention of tenant data
+
+## Problem
+
+In Cortex, we are missing retention of blocks storage. The metric data is growing over time per-tenant, at the same time, the value of data decreases. We want to support retention time of metric data like prometheus does.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:16]: This is not entire correct. Data retention is typically achieved be achieved via a bucket policy. However, we should explain this has two main issues:

1. Not every backend storage support bucket policies
2. Bucket policies don't easily allow a per-tenant custom retention

File: docs/proposals/tenant_retention.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.
+
+## Proposal
+
+### API Endpoints
+
+#### POST /compactor/retention_tenant
+
+We propose to introduce an `/compactor/retention_tenant` API endpoint to set data retention for the tenant.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:28]: I wouldn't introduce any new API. We should use the already existing mechanism used to set per-tenant overrides (via the runtime config "overrides").

File: docs/proposals/tenant_retention.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+To query the state of data retention, an endpoint will be available: `/compactor/retention_tenant`.
+This will return the retention of current tenant.
+For consistency with other endpoints this endpoint would therefore require an `X-Scope-OrgID` header and use the tenant from it.
+
+### Implementation (asynchronous)
+
+Compactor will implement both API endpoints.
+Upon receiving the call to `/compactor/retention_tenant` endpoint, Compactor will initiate the retention by writing “retetion marker” objects for specific tenant to blocks bucket.
+
+Retention marker for the tenant will be an object stored under tenant prefix, eg. “/<tenant>/retention”.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:46]: I wouldn't use any retention marker. The per-tenant retention should be part of the config.

**Review by pracucci — COMMENTED** [review]
Thanks for working on this proposal! I left a couple of high level comments. I would love going deeper into the review once these are addressed. Thanks!

**Comment by AllenzhLi** [comment]
Sorry, I haven't paid attention to it for a while and got stuck，now I'm back.  

I have test the runtime_config, it is looks nice and simplify except that I can't update the setting via a http api.

**Comment by AllenzhLi** [comment]
#3622

**Comment by pstibrany** [comment]
> I have test the runtime_config, it is looks nice and simplify except that I can't update the setting via a http api.

There is no API to change runtime config. You can modify YAML file directly, and Cortex will reload it automatically.

**Comment by AllenzhLi** [comment]
> There is no API to change runtime config. You can modify YAML file directly, and Cortex will reload it automatically.

Yeah，it will reload periodly. 

But update the retention duration onfly is also needed.

https://grafana.com/docs/loki/latest/operations/storage/retention/
Loki: Currently, the retention policy can only be set globally. A per-tenant retention policy with an API to delete ingested logs is still under development.

**Comment by pstibrany** [comment]
> Yeah，it will reload periodly.
> 
> But update the retention duration onfly is also needed.
> 
> https://grafana.com/docs/loki/latest/operations/storage/retention/
> Loki: Currently, the retention policy can only be set globally. A per-tenant retention policy with an API to delete ingested logs is still under development.

You are right, but that is a separate problem from retention. I'd suggest focusing on retention in this design doc, with options we already have (runtime config), and leave solution for API for a different design document. (Which then also needs to discuss where to store such modified options to survive restarts).

**Comment by AllenzhLi** [comment]
> You are right, but that is a separate problem from retention. I'd suggest focusing on retention in this design doc, with options we already have (runtime config), and leave solution for API for a different design document. (Which then also needs to discuss where to store such modified options to survive restarts).

I agree with you. I will update the design document as soon as possible.

File: docs/proposals/tenant_retention.md (hunk 1 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: I agree with you. I will update this.

File: docs/proposals/tenant_retention.md (hunk 2 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: I will delete the api design.

File: docs/proposals/tenant_retention.md (hunk 3 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: I agree with you, this will be delete.

File: docs/proposals/tenant_retention.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.
+
+### runtime config
+Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.
+
+## Proposal
+
+### retention component
+
+We propose to introduce an new runtime config component `retention`.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:33]: I don't think this is necessary. We can just have 1 new field in the `Limits` struct (defined at `pkg/util/validation/limits.go`) which allows to specify the per-tenant retention (0 to disable it).

File: docs/proposals/tenant_retention.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+Per-tenant retention configuration will be applyed for in follow order:
+
+* tenant specified value
+* global value
+
+For each tenant, if a tenant-specific value exists, it will be used directly, otherwise, if a global value exists, then the global value will be used; If neither exists, do nothing.
+
+### Implementation
+
+The runtime config will add a retention component which include a dictionary to store per-tenant retention value and global retention value.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:46]: See above. No need for a new dictionary/map. We can just reuse `Limits`, which is the way we currently implement any per-tenant config.

File: docs/proposals/tenant_retention.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+Compactor will check per-tanant retention value and compare to meta info of blocks. The blocks that match `maxTime < now - retention` will be marked for delete.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:48]: FYI, we have a blocks cleaner component within the compactor. That could be the right place to add the retention logic.

File: docs/proposals/tenant_retention.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+## Compactor
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:50]: I'm not sure what's the purpose of this section. Is it to explain how things are working today in the compactor?

**Review by pracucci — COMMENTED** [review]
Thanks for the follow up! The updated proposal LGTM, modulo few last comments.

File: docs/proposals/tenant_retention.md (hunk 4 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: You are right.  I didn't understand the runtime config correctly.

File: docs/proposals/tenant_retention.md (hunk 5 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: Same as above. I will reuse the `Limits`.

File: docs/proposals/tenant_retention.md (hunk 6 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: I agree with you. I tend to insert the logic in function `cleanUser`.

File: docs/proposals/tenant_retention.md (hunk 7 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md]: I want to explain how to implement blocks delete.

File: docs/proposals/tenant_retention.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.
+
+### runtime config
+Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See [runtime-configuration-file](https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file) for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.
+
+## Proposal
+
+### retention period field
+
+We propose to introduce just one new field `RetentionPeriodPerTenant` in the Limits struct(defined at pkg/util/validation/limits.go).
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:33]: All fields in the `Limits` struct are per-tenant so no need to make it explicit.

```suggestion
We propose to introduce just one new field `RetentionPeriod` in the Limits struct(defined at pkg/util/validation/limits.go).
```

File: docs/proposals/tenant_retention.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+`RetentionPeriodPerTenant` setting how long historical metric data retention period per-tenant. `0` is disable.
+
+As a runtime config, the settings is reload period (which defaults to 10 seconds), so we can update the settings onfly.
```
Comment (pracucci) [inline docs/proposals/tenant_retention.md:37]:
```suggestion
Runtime config is reloaded periodically (defaults to 10 seconds), so we can update the retention settings on-the-fly.
```

**Review by pracucci — APPROVED** [review]
The proposal LGTM, thanks! I left a couple of non-blocking nits.

@gouthamve @pstibrany @ranton256 (which was interested as well) could you take a look, please?

File: docs/proposals/tenant_retention.md [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+`RetentionPeriodPerTenant` setting how long historical metric data retention period per-tenant. `0` is disable.
+
+Runtime config is reloaded periodically (defaults to 10 seconds), so we can update the retention settings on-the-fly.
+
+For each tenant, if a tenant-specific *runtime_config* value exists, it will be used directly, otherwise, if a default *limits_config* value exists, then the default value will be used; If neither exists, do nothing.
+
+### Implementation
+
+A BlocksCleaner within the Compactor run periodically(which defaults to 1 hour) and the retention logic will insert into it. The logic should compare retention value to block `maxTime` and blocks that match `maxTime < now - retention` will be marked for delete.
```
Comment (pstibrany) [inline docs/proposals/tenant_retention.md:43]:
```suggestion
A BlocksCleaner within the Compactor run periodically (which defaults to 15 minutes) and the retention logic will insert into it. The logic should compare retention value to block `maxTime` and blocks that match `maxTime < now - retention` will be marked for delete.
```

File: docs/proposals/tenant_retention.md [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+### runtime config
+Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See [runtime-configuration-file](https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file) for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.
+
+## Proposal
+
+### retention period field
+
+We propose to introduce just one new field `RetentionPeriod` in the Limits struct(defined at pkg/util/validation/limits.go).
+
+`RetentionPeriodPerTenant` setting how long historical metric data retention period per-tenant. `0` is disable.
```
Comment (pstibrany) [inline docs/proposals/tenant_retention.md:35]: I assume `RetentionPeriod` and `RetentionPeriodPerTenant` are the same. We should use just one name here.

**Review by pstibrany — APPROVED** [review]
LGTM, thank you!

File: docs/proposals/tenant_retention.md (hunk 11 above)
Comment (AllenzhLi) [reply docs/proposals/tenant_retention.md:35]: Yes, they are same. Both should be `RetentionPeriod`.

**Comment by pracucci** [comment]
This design doc looks solid to me. Simple yet effective. 👍

**Review by ranton256 — APPROVED** [review]
Thanks for writing this up. I concur with other reviews, this is straightforward to address the problem reusing the existing override mechanism and Compactor component, LGTM.

**Pull request merged**